### DISCUSSION TITLE
How do we write to the log?

### DISCUSSION THREAD

**Discussion by PPetterssonSO** [SuperOffice/typescript-for-superoffice#26] [General]
Since tracing have a lovley tendency to disable itself on crashes, we sometimes need to use the log("whatever");
How can we achieve this in TypeScript?

**Comment by ejfasting** [comment]
You could use the DatabaseTableAgent to insert into `log_events`. There is no helper-method to do it for you, at least not right now.

**Reply by PPetterssonSO** [reply]
Aah perfect, thank you!

**Reply by ejfasting** [reply]
You can take inspiration from the [log helper method](https://docs.superoffice.com/en/automation/crmscript/debugging/helper-functions-for-logging.html) for crmscript, to create a helper-method useable in typescript :)

**Comment by PPetterssonSO** [comment]
Sorry for bringing back an closed request, but, im 100% stuck on the log script and I need help :)

This is the code:
`async function log(value: string) {
  var databaseTableAgent = soApi.getDatabaseTableAgent();
  const values: { [key: string]: string } = {};
  values["var1"] = value;
  values["var2"] = "";
  values["time_of_event"] = "2025-01-02 08:46:56";
  values["event_type"] = "1";
  values["event_group"] = "13";
  values["target_id"] = "-1";

  const json = {
    tableName: "log_events",
    values: values,
  };

  try {
    return await databaseTableAgent.insertRowAsync(json);
  }
  catch (error) {
    printLine(error);
  }
}`

The json parameter seems to be valid. 
{"tableName":"log_events","values":{"var1":"tjena","var2":"","time_of_event":"2025-01-02 08:46:56","event_type":"1","event_group":"13","target_id":"-1"}

But, im getting the following error: AxiosError: Request failed with status code 403

What in the gods name am I missing? :)

**Reply by ejfasting** [reply]
Hi @Nawksen, 

Sorry, it just occurred to me that you cant write to 'default tables' with the databasetableagent, only extra_tables are allowed.. My bad!
This is something we need to add functionality for, then, and I'm making a backlog item to either add this as a 'special function/method' you can use, OR add `log_events` to the whitelist for databasetableagent.

**Reply by ejfasting** [reply]
A possible workaround would be to create an extraTable, x_log, and write to that instead. We havent fully decided what we will move forward with, as the DatabaseTableAgent are restricted to only handle ExtraTables and ExtraFields (just looked at the code). 
Will have to update this thread at a later point

**Comment by ejfasting** [comment]
Hello @Nawksen, 

Just a quick update on this. We have added a new method for this, which is located on `context`.
Developer notes:
`
Added context.log("log messsage"), which should log in the same way before. A couple of changes though:
We do not log the pid anymore
We log the associate id and name of the user in source_user and source_user_desc 
`

Hope this is a practical solution for you, and let me know your thoughts when you give it a spin! :)

//Eivind
